File: 241017/안전 지대/comfort-zone.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
#define MAX_NM 50
#define DIR_NUM 4

int n, m;
int grid[MAX_NM][MAX_NM];
bool visited[MAX_NM][MAX_NM];

void Init_visit() {
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            visited[i][j] = false;
        }
    }
}

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<m;
}

bool Cango(int x, int y, int k) {
    if(!InRange(x, y)) return false;
    if(visited[x][y] || grid[x][y] <= k) return false;

    return true;
}

void DFS(int x, int y, int k) {
    int dx[DIR_NUM] = {-1, 0, 1, 0};
    int dy[DIR_NUM] = {0, -1, 0, 1};

    for(int d=0; d<DIR_NUM; d++) {
        int nx = x+dx[d];
        int ny = y+dy[d];

        if(Cango(nx, ny, k)) {
            visited[nx][ny] = true;
            DFS(nx, ny, k);
        }
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >>n>>m;

    int max_k = 0;
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin >> grid[i][j];
            max_k = max(max_k, grid[i][j]);
        }
    }

    int max_saf = -1;
    int ans = 0;
    for(int k=1; k<=max_k; k++) {
        Init_visit();
        int cnt = 0;
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(grid[i][j] > k && visited[i][j] == false) {
                    cnt++;
                    visited[i][j] = true;
                    DFS(i, j, k);
                }
            }
        }

        // cout << "------" << cnt << '\n';

        if(max_saf < cnt) {
            max_saf = cnt;
            ans = k;
        }
    }

    cout << ans << ' ' << max_saf;
    return 0;
}

File: 241031/안전 지대/comfort-zone.py
n, m = tuple(map(int, input().split()))

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]


def inRange(x, y):
    return 0<=x and x<n and 0<=y and y<m

def canGo(x, y, k, visited):
    if not inRange(x, y):
        return False
    if grid[x][y] <= k or visited[x][y]:
        return False
    
    return True


def dfs(x, y, k, visited):
    dxs, dys = [-1, 0, 1, 0], [0, 1, 0, -1]

    for dx, dy in zip(dxs, dys):
        nx, ny = x+dx, y+dy

        if canGo(nx, ny, k, visited):
            visited[nx][ny] = 1
            dfs(nx, ny, k, visited)


ans = 0
ans_k = 0

max_k = max(max(row) for row in grid)

for k in range(max_k+1):
    cnt = 0
    visited = [[0 for _ in range(m)] for _ in range(n)]
    for i in range(n):
        for j in range(m):
            if canGo(i, j, k, visited):
                cnt += 1
                visited[i][j] = 1
                dfs(i, j, k, visited)

    if ans < cnt:
        ans = cnt
        ans_k = k


print(ans, ans_k)